File: ProblemSolve/316.cpp
/****************************
 * Remove Duplicate Letters *
 ****************************/

class Solution {
public:
    string removeDuplicateLetters(string s) {
        vector<int> candidate(256);
        vector<bool> visitied(256);
        
        for(char ch : s){
            candidate[(int)ch]++;
        }
        
        string ans = "0";
        for(char ch : s){
            candidate[(int)ch]--;
            
            if(visitied[ch]){
                continue;
            }
            
            while(ch < ans.back() && candidate[ans.back()] != 0){
                visitied[ans.back()] = false;
                ans.pop_back();
            }
            
            ans += ch;
            visitied[ch] = true;
        }
        
        return ans.substr(1);
    }
};

File: ProblemSolve/316.cs
/****************************
 * Remove Duplicate Letters *
 ****************************/

public class Solution {
    public int Back(StringBuilder sb){
        return (int)sb[sb.Length - 1];
    }
    
    public void PopBack(ref StringBuilder sb){
        sb.Remove(sb.Length-1, 1);
    }
    
    public string RemoveDuplicateLetters(string s) {
        int[] candidate = new int[256];
        bool[] visitied = new bool[256];
        
        foreach(char ch in s){
            candidate[(int)ch]++;
        }
        
        StringBuilder sb = new StringBuilder();
        sb.Append('0');
        
        foreach(char ch in s){
            candidate[(int)ch]--;
            
            if(visitied[(int)ch]){
                continue;
            }
            
            while((int)ch < Back(sb) && candidate[Back(sb)] > 0){
                visitied[Back(sb)] = false;
                PopBack(ref sb);
            }
            
            sb.Append(ch);
            visitied[(int)ch] = true;
        }
        sb.Remove(0,1);
        
        return sb.ToString();
        
    }
}